sound/hybrid_font.h

```diff
@@ -450,9 +450,9 @@ class HybridFont : public SaberBase, public Looper {
           size_t total = SFX_poweroff.files_found() + SFX_pwroff.files_found();
           if (total) {
             if ((rand() % total) < SFX_poweroff.files_found()) {
-              PlayMonophonic(&SFX_poweroff, NULL);
+              PlayCommon(&SFX_poweroff);
```

Comment:
Won't this make monophonic fonts keep playing the hum after poweroff?

Reply:
Yes. Yes it does :(
My test was inadequate.
Maybe something with the check_postoff_ then based on whether monophonic_hum_ or not?
This part:
https://github.com/profezzorn/ProffieOS/blob/dee01ebe43d24bf469521307033d640beb75f720/sound/hybrid_font.h#L469

Reply:
I'm not sure I understand the problem.
When does this actually occur?
CFX fonts with humm?
Is there a bug report or thread I've forgotten about?

Reply:
Sorry, yes, this is issue #159 I thought I'd take a crack at it.

I just added a different change and it does test ok (used an actual monophonic font this time).
Being honest, I am poking around in the dark as I don't yet 100% grasp the way mono/poly/looped sounds work. 
I think the NULL I replaced is the place where 'loop' goes .... but sticking pstoff in there seems to work.
Again, somewhat blindly stabbing at this, but trying to help.

Reply:
> CFX fonts with humm?

And actually these too, yes. Fonts with humM are polyphonic I believe, and interestingly are not affected by the change here, whereas the "legacy" monophonic one does respond positively and plays pstoff.

Some insight as to how pstoff is getting called here would be appreciated, as I still am rather green understanding even the functions PlayMonophonic/Polyphonic themselves. The latter is set as a RefPtr... 
As much as I examine the code, it's tough to follow/swallow.

Reply:
So, are you sure there is actually a problem here?
I looked at the code, and it seems like it should just work.
poweroff, pwroff and "in" are all linked to pstoff in the same way.

Reply:
ok. Triple checked and narrowed some of it down.
It does seem that the humM one is the issue, not hum.
Example below, both have pstoff files. 
NIGHTFALL has hum and is a monophonic font.
postoff shows playing AND is audible after either poweroff or pwroff.

ANCIENT_PROPHECIES has humM and is polyphonic.
SOMETIMES pstoff is shown in serial monitor as playing, sometimes not, and in either case, and with either poweroff or pwroff, the audio is NOT played, although it says it should be.


```
21:37:37.222 -> Ignition.
21:37:37.257 -> unit = 0 vol = 0.50, Playing 1-NIGHTFALL/poweron.wav
21:37:37.257 -> channels: 1 rate: 22050 bits: 16
21:37:37.257 -> unit = 1 vol = 0.00, Playing 1-NIGHTFALL/lswing2.wav
21:37:37.257 -> channels: 1 rate: 22050 bits: 16
21:37:37.257 -> unit = 2 vol = 0.00, Playing 1-NIGHTFALL/hswing2.wav
21:37:37.291 -> channels: 1 rate: 22050 bits: 16
21:37:37.291 -> Motion setup ... done.
21:37:37.325 -> unit = 3 vol = 0.50, Playing 1-NIGHTFALL/swing3.wav
21:37:37.325 -> channels: 1 rate: 22050 bits: 16
21:37:37.325 -> EVENT: Swing ON millis=325915
21:37:37.325 -> Audio underflows: 3
21:37:38.094 -> Playing 1-NIGHTFALL/hum.wav
21:37:38.094 -> channels: 1 rate: 22050 bits: 16
21:37:44.263 -> unit = 0 vol = 0.50, Playing 1-NIGHTFALL/poweroff.wav
21:37:44.263 -> channels: 1 rate: 22050 bits: 16
21:37:44.943 -> Playing 1-NIGHTFALL/pstoff.wav
21:37:44.943 -> channels: 1 rate: 44100 bits: 16
21:37:44.943 -> Audio underflows: 2
21:37:46.625 -> Unmounting SD Card.
21:37:46.625 -> Amplifier off.

21:39:33.541 -> Ignition.
21:39:33.541 -> unit = 3 vol = 0.50, Playing 1-NIGHTFALL/poweron3.wav
21:39:33.541 -> channels: 1 rate: 22050 bits: 16
21:39:33.541 -> unit = 2 vol = 0.00, Playing 1-NIGHTFALL/lswing2.wav
21:39:33.541 -> channels: 1 rate: 22050 bits: 16
21:39:33.541 -> unit = 1 vol = 0.00, Playing 1-NIGHTFALL/hswing2.wav
21:39:33.580 -> channels: 1 rate: 22050 bits: 16
21:39:35.466 -> Playing 1-NIGHTFALL/hum.wav
21:39:35.466 -> channels: 1 rate: 22050 bits: 16
21:39:36.375 -> unit = 0 vol = 0.50, Playing 1-NIGHTFALL/pwroff2.wav
21:39:36.375 -> channels: 1 rate: 22050 bits: 16
21:39:36.798 -> Playing 1-NIGHTFALL/pstoff.wav
21:39:36.798 -> channels: 1 rate: 44100 bits: 16
21:39:36.798 -> Audio underflows: 3
21:39:38.459 -> Unmounting SD Card.
21:39:38.492 -> Amplifier off.



21:40:38.513 -> Ignition.
21:40:38.546 -> unit = 0 vol = 0.50, Playing 17-ANCIENT_PROPHECIES/preon1.wav
21:40:38.546 -> channels: 1 rate: 44100 bits: 16
21:40:38.546 -> Motion setup ... done.
21:40:39.461 -> Playing 17-ANCIENT_PROPHECIES/poweron2.wav
21:40:39.461 -> channels: 1 rate: 44100 bits: 16
21:40:39.461 -> unit = 1 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/humm1.wav
21:40:39.461 -> channels: 1 rate: 44100 bits: 16
21:40:39.461 -> unit = 2 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/lswing4.wav
21:40:39.496 -> channels: 1 rate: 44100 bits: 16
21:40:39.496 -> unit = 3 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/hswing4.wav
21:40:39.496 -> channels: 1 rate: 44100 bits: 16
21:40:43.745 -> unit = 0 vol = 0.50, Playing 17-ANCIENT_PROPHECIES/poweroff.wav
21:40:43.745 -> channels: 1 rate: 44100 bits: 16
21:40:45.656 -> Playing 17-ANCIENT_PROPHECIES/pstoff.wav
// *********** shown as played but did not
21:40:45.656 -> channels: 1 rate: 44100 bits: 16
21:40:45.656 -> Audio underflows: 1
21:40:45.656 -> Unmounting SD Card.
21:40:45.689 -> Amplifier off.
21:40:51.450 -> Battery voltage: 3.58
21:40:51.980 -> Ignition.
21:40:52.016 -> unit = 0 vol = 0.50, Playing 17-ANCIENT_PROPHECIES/preon1.wav
21:40:52.016 -> channels: 1 rate: 44100 bits: 16
21:40:52.955 -> Playing 17-ANCIENT_PROPHECIES/poweron.wav
21:40:52.955 -> channels: 1 rate: 44100 bits: 16
21:40:52.955 -> unit = 1 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/humm1.wav
21:40:52.955 -> channels: 1 rate: 44100 bits: 16
21:40:52.955 -> unit = 2 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/lswing1.wav
21:40:52.955 -> channels: 1 rate: 44100 bits: 16
21:40:52.955 -> unit = 3 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/hswing1.wav
21:40:52.989 -> channels: 1 rate: 44100 bits: 16
21:40:56.197 -> unit = 0 vol = 0.50, Playing 17-ANCIENT_PROPHECIES/pwroff2.wav
21:40:56.197 -> channels: 1 rate: 44100 bits: 16
// *********** no pstoff shown as played and also not played.
21:40:58.103 -> Unmounting SD Card.
21:40:58.103 -> Amplifier off.
21:41:00.266 -> Ignition.
21:41:00.303 -> unit = 0 vol = 0.50, Playing 17-ANCIENT_PROPHECIES/preon1.wav
21:41:00.303 -> channels: 1 rate: 44100 bits: 16
21:41:01.219 -> Playing 17-ANCIENT_PROPHECIES/poweron2.wav
21:41:01.219 -> channels: 1 rate: 44100 bits: 16
21:41:01.219 -> unit = 1 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/humm1.wav
21:41:01.253 -> channels: 1 rate: 44100 bits: 16
21:41:01.253 -> unit = 3 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/lswing4.wav
21:41:01.253 -> channels: 1 rate: 44100 bits: 16
21:41:01.253 -> unit = 2 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/hswing4.wav
21:41:01.289 -> channels: 1 rate: 44100 bits: 16
21:41:03.658 -> unit = 0 vol = 0.50, Playing 17-ANCIENT_PROPHECIES/poweroff.wav
21:41:03.693 -> channels: 1 rate: 44100 bits: 16
21:41:05.582 -> Playing 17-ANCIENT_PROPHECIES/pstoff.wav
21:41:05.582 -> channels: 1 rate: 44100 bits: 16
21:41:05.582 -> Audio underflows: 2
21:41:05.582 -> Unmounting SD Card.
21:41:05.616 -> Amplifier off.
21:41:07.709 -> Ignition.
21:41:07.747 -> unit = 0 vol = 0.50, Playing 17-ANCIENT_PROPHECIES/preon1.wav
21:41:07.747 -> channels: 1 rate: 44100 bits: 16
21:41:08.661 -> Playing 17-ANCIENT_PROPHECIES/poweron4.wav
21:41:08.661 -> channels: 1 rate: 44100 bits: 16
21:41:08.661 -> unit = 1 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/humm1.wav
21:41:08.661 -> channels: 1 rate: 44100 bits: 16
21:41:08.661 -> unit = 3 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/lswing1.wav
21:41:08.695 -> channels: 1 rate: 44100 bits: 16
21:41:08.695 -> unit = 2 vol = 0.00, Playing 17-ANCIENT_PROPHECIES/hswing1.wav
21:41:08.695 -> channels: 1 rate: 44100 bits: 16
21:41:10.718 -> unit = 0 vol = 0.50, Playing 17-ANCIENT_PROPHECIES/pwroff2.wav
21:41:10.754 -> channels: 1 rate: 44100 bits: 16
21:41:11.427 -> Battery voltage: 3.57
21:41:12.624 -> Playing 17-ANCIENT_PROPHECIES/pstoff.wav
// *********** shown as played but did not
21:41:12.660 -> Audio underflows: 1
21:41:12.660 -> Unmounting SD Card.
21:41:12.660 -> Amplifier off.
```

Reply:
I tihnk the problem is this line:
https://github.com/profezzorn/ProffieOS/blob/a17f78bec777fc57f094091e0e4c014a7979731b/sound/hybrid_font.h#L469

The code unconditionally plays poweroff/pwroff the monophonic way, but then it sets state_ differently depending on weather the hum is monophonic or polyphonic. This would be correct only if the playback was different.

Either the way we do playback is wrong, or the way we set state_ is wrong.

Is poweroff mixed with the hum in polyphonic CFX fonts?

Reply:
That's the line I looked at before as well.

> Is poweroff mixed with the hum in polyphonic CFX fonts?

It is not in the examples I have from the default package downloaded from Plecter "CrystalFocus-V10.006H-DefaultPackage-SMOOTHED.zip"
Seems to be they are true polyphonic if it is humM.

Reply:
In that case, I think I know how it should work and why pstoff doesn't work.
I checked in a fix that plays poweroff polyphonically if humm is available.
By having state_ match the play method, I think pstoff should also work.
Please let me know how it works for you.